src/transports/tcp.rs

```diff
@@ -473,35 +508,38 @@ impl<Id: PeerId> Transport<Id> for TcpTransport<Id> {
                     PeerNetError::SendError.error("error set write timeout", Some(e.to_string()))
                 })?;
 
-            endpoint
-                .stream
-                .write_all(&msg_size.to_be_bytes())
-                .map_err(|err| {
-                    TcpError::ConnectionError.wrap().new(
+            match endpoint.stream.write(&msg_size.to_be_bytes()) {
+                Ok(0) => {
+                    endpoint.shutdown();
+                    return Err(TcpError::ConnectionError
+                        .wrap()
+                        .error("send len write = 0", None));
+                }
+                Ok(_count) => {}
+                Err(err) => {
+                    return Err(TcpError::ConnectionError.wrap().new(
                         "send len write",
                         err,
                         Some(format!("{:?}", data.len().to_le_bytes())),
-                    )
-                })?;
+                    ))
+                }
+            }
 
             let mut total_bytes_written = 0;
-            let chunk_size = 1024;
-
             while total_bytes_written < data.len() {
```

Comment:
same remark as above, just call `write_exact_timeout` instead of expanding the logic here as well